include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -97,41 +97,140 @@ template <typename> class FullMatrix;
+   * vmult() function, but operating on plain pointers rather than a vector
+   * (no check of array bounds possible).
+   */
+  void vmult (Number *dst,
+              const Number *src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * apply_inverse() function, but operating on plain pointers rather than a
+   * vector (no check of array bounds possible).
+   */
+  void apply_inverse (Number *dst,
+                      const Number *src) const;
+
+protected:
+  /**
+   * Constructor.
+   */
+  TensorProductMatrixSymmetricSumBase () = default ;
+
+  /**
+   * A copy of the @p mass_matrix object passed to the reinit() method.
+   */
+  std::array<Table<2,Number>,dim> mass_matrix;
+
+  /**
+   * A copy of the @p derivative_matrix object passed to the reinit() method.
+   */
+  std::array<Table<2,Number>,dim> derivative_matrix;
+
+  /**
+   * A vector containing the generalized eigenvalues of A s = lambda B s.
+   */
+  std::array<AlignedVector<Number>,dim> eigenvalues;
+
+  /**
+   * The matrix containing the generalized eigenvectors.
+   */
+  std::array<Table<2,Number>,dim> eigenvectors;
+
+private:
+  /**
+  * An array for temporary data.
+  */
+  mutable AlignedVector<Number> tmp_array;
+
+  /**
+   * A mutex that guards access to the array @p tmp_array.
+   */
+  mutable Threads::Mutex mutex;
+};
+
+
+
+/**
+ * ... new TensorProductMatrixSymmetricSum using the base class as tensor product
+ * container and interface to arithmetic operations for a generic Number type ...
+ */
+template <int dim, typename Number, int size = -1>
 class TensorProductMatrixSymmetricSum
+  : public TensorProductMatrixSymmetricSumBase<dim,Number,size>
 {
 public:
   /**
    * Constructor.
    */
-  TensorProductMatrixSymmetricSum();
+  TensorProductMatrixSymmetricSum () ;
 
   /**
    * Constructor that is equivalent to the previous constructor and
-   * immediately calling reinit().
+   * immediately calling the corresponding reinit().
    */
-  TensorProductMatrixSymmetricSum(const FullMatrix<Number> &mass_matrix,
-                                  const FullMatrix<Number> &derivative_matrix);
+  TensorProductMatrixSymmetricSum (const std::array<Table<2,Number>, dim> &mass_matrix,
+                                   const std::array<Table<2,Number>, dim> &derivative_matrix) ;
 
   /**
-   * Initializes the matrix to the given mass matrix $M$ and derivative matrix
-   * $A$. Note that the current implementation requires $M$ to be symmetric
-   * and positive definite and $A$ to be symmetric and invertible but not
+   * Constructor that is equivalent to the first constructor and
+   * immediately calling the corresponding reinit().
+   */
+  TensorProductMatrixSymmetricSum (const std::array<FullMatrix<Number>,dim> &mass_matrix,
+                                   const std::array<FullMatrix<Number>,dim> &derivative_matrix) ;
```

Comment:
Do you think that it still makes sense to keep the arguments `FullMatrix` types? I'd rather go with `Table<2,Number>`. On the one hand, that interface was introduced less than two months ago, and `FullMatrix` is also a `Table`, so it'd stay compatible. Or why did you have to add both constructors here?

Reply:
I totally agree with you. My intent was to present a first version that is compatible with the old one and mapping_q_generic.cc uses the `TensorProductMatrixSymmetricSum` initialized by means of `FullMatrix` arguments. I did not want to change any of the existing code (outside of tensor_product_matrix.h) before discussing this with you. I think we should only have a interface for `Table<2,Number>`.

Reply:
I tried to avoid this constructor/`reinit`, however, the compiler does not (implicitly) convert a `std::array<FullMatrix>` into a `std::array<Table>`.